src/ros_can_nodes/src/RosCanNode.cpp

```diff
@@ -239,69 +243,48 @@ namespace roscan {
-        RosIntrospection::RenamedValues& renamed_values = renamed_vectors[topic_name];
-
-        //copy raw memory into the buffer
-        buffer.resize(msg->size());
-        ros::serialization::OStream stream(buffer.data(), buffer.size());
-        msg->write(stream);
-
-        //deserialize and rename the vectors
-        parser.deserializeIntoFlatContainer(topic_name, absl::Span<uint8_t>(buffer), &flat_container, 100);
+    void RosCanNode::rosCanCallback(const ShapeShifter::ConstPtr& msg, const uint8_t topicID, const std::string& topic_name) {
+        ROS_INFO("callback: topic_id = %d topic_name = %s", topicID, topic_name.c_str());
 
-        parser.applyNameTransform(topic_name, flat_container, &renamed_values);
+        IntrospectionHelpers::registerMessage(msg, topic_name);
+        const auto buf = IntrospectionHelpers::modify_buffer(msg->getDataType(), msg->raw_data(), msg->size());
+        const auto msg_count = buf.size() / 8u + (buf.size() % 8u != 0u);
+        ROS_INFO("buf size %d msg_count %d", buf.size(), msg_count);
 
-        //print the content of the message
-        //printf("--------- %s ----------\n", topic_name.c_str());
-        for (const auto& it: renamed_values) {
-            const std::string& key = it.first;
-            const RosIntrospection::Variant& value   = it.second;
-            //printf(" %s = %f\n", key.c_str(), value.convert<double>()); //convert into CAN message set
-        }
-
-        canid_t header = 0x0;
+        canid_t header = 0;
+        header |= CAN_EFF_FLAG;
 
-        printf("can topic id 0x%x\n", topicID);
+        ROS_INFO("can topic id 0x%x", topicID);
         header |= (1 << ROSCANConstants::Common::bitshift_mode);
         header |= (0 << ROSCANConstants::Common::bitshift_priority);
         header |= (0 << ROSCANConstants::Common::bitshift_func);
         header |= (((uint32_t)topicID) << ROSCANConstants::ROSTopic::bitshift_topic_id);
         header |= (0 << ROSCANConstants::Common::bitshift_seq);
-        printf("topic id at position 0x%x\n", (((uint32_t)topicID) << ROSCANConstants::ROSTopic::bitshift_topic_id));
+        ROS_INFO("topic id at position 0x%x", (((uint32_t)topicID) << ROSCANConstants::ROSTopic::bitshift_topic_id));
 	    header |= (id_ << ROSCANConstants::ROSTopic::bitshift_nid);
         header |= (((msg_num = ((msg_num + 1)  % 3)) & ROSCANConstants::ROSTopic::bitmask_msg_num) << ROSCANConstants::ROSTopic::bitshift_msg_num);
         header |= (1 << ROSCANConstants::ROSTopic::bitshift_len);
         header |= CAN_EFF_FLAG;
 
-        can_frame frame;
-
-        frame.can_id = header;
-        frame.can_dlc = 8;
-
-        for (const auto& it: renamed_values) {
-            *(double *)frame.data = it.second.convert<double>();
+        for (auto i = 0u;i < msg_count;++i) {
+            can_frame frame;
+            frame.can_id = header;
+
+            frame.can_id |= (i & 0b11) << ROSCANConstants::ROSTopic::bitshift_msg_num;
```

Comment:
msg_num is a per message thing, not per packet. It should never change in this loop